### ISSUE TITLE
Example with redux-thunk

### ISSUE THREAD

**Issue by marr** [ryanseddon/redux-debounced#4]
Hi, I was hoping to have an example of this middleware in addition to using async actions with `redux-thunk`. I have an actions export like this:

```
export function asyncFindUsers(search) {
  console.log('asyncFindUsers')
  return (dispatch) => {
    let users = []; // TODO: Add fetch from API
    setTimeout(() => {
      return dispatch(findUsers(users));
    }, 500)
  };
}

function findUsers(users) {
  return { type: FIND_USERS, users };
}
```

And the wiring inside the view:

```
function mapDispatchToProps(dispatch) {
  return {
    changeRoute: (url) => dispatch(routeActions.push(url)),
    findUsers: (value) => dispatch(asyncFindUsers(value)),
  };
}
```

The docs suggest using the `meta` property in the action that's returned, but since thats async when using `redux-thunk` I think there may be additional steps required to debounce the action.

**Comment by Anahkiasen** [comment]
:+1: Wondering also how to use with `redux-thunk`

**Comment by seansfkelley** [comment]
If you put the debounce middleware before the thunk middleware, you can just attach `meta` to the function object:

```
function action() {
  const thunk = (dispatch, getState) => { ... };
  thunk.meta = {
    debounce: {
      time: 100,
      key: 'my-thunk-action'
    }
  };
  return thunk;
}
```

**Comment by mschipperheyn** [comment]
Does this also work for `return async dispatch =>` es 7 decorator style functions?

**Comment by blessanm86** [comment]
Does this still work? Cause I tried something like below and it logs continuously.

```
export function trackCustomerSearch(key) {
  const thunk = dispatch => {
    console.log('Search Key ----> ', key);  
  };
  thunk.meta = {
    debounce: {
      time: 2500
    }
  };
  return thunk;
}
```

**Comment by seansfkelley** [comment]
You have to use a key, since a thunk has no type. You should also ensure you put the debounce middleware before the thunk middleware.

**Comment by ryanseddon** [comment]
Once you get this working it'd be great to do a PR to document what needs to be done so someone else doesn't get stuck too.

**Comment by blessanm86** [comment]
@seansfkelley Thanks, the key was exactly the issue. I wasn't sure whether I needed it. I can look into adding this into the docs this weekend.

**Comment by krstffr** [comment]
If you're just looking for a way of using lodash's debounce function in your code, this is an easier way of debouncing thunk actions (according to me): https://gist.github.com/krstffr/245fe83885b597aabaf06348220c2fe9

**Comment by seansfkelley** [comment]
The only downsides to your method, @krstffr, is that proper testing becomes wonky (as you now have to deal with timings) and you cannot share the action across multiple store instances (which is what drove me to use this middleware in the first place).

**Comment by ropez** [comment]
Unfortunately, the described method breaks one use-case: If you want to create a "simple" debounced action inside a thunk, the debounce middleware needs to be applied after the tunk middleware, or it doesn't work:

```
const simpleDebouncedAction = () => ({
    type: 'SIMPLE_ACTION',
    meta: {
        debounce: { time: 1000 }
    }
})

const asyncAction = () => dispatch => {
    setTimeout(() => {
        dispatch(simpleDebouncedAction())
    }, 10000)
}
```

This only works if the debounced middleware is applied _after_ the thunk middleware. (I tried to have it both before and after, but that had the undesired side-effect of "double-debouncing" simple actions that doesn't involve thunk)

**Comment by ropez** [comment]
Nevermind the previous comment. It was my fault. It was because I did this:

```
createStore(reducer, compose(
  applyMiddleware(createDebouncer()),
  applyMiddleware(thunk),
  <some other enhancers>
))
```

I fixed the issue by doing this:

```
createStore(reducer, compose(
  applyMiddleware(
    createDebouncer(),
    thunk
  ),
  <some other enhancers>
))
```

**Comment by tswaters** [comment]
I'm having a problem chaining the results of dispatch.

According to redux-thunk docs, one should be able to call `then` on on the dispatch function, 
e.g., 

```js
dispatch({action, type}).then(() => {/*it resolved!*/})
```

With this middleware in the mix, the dispatch function is returning undefined instead of a promise, so the call fails.  If I pull out the `meta` value from the action, I'm able to call `then` on the dispatch as expected.

I tried modifying the `lib/index.js` directly in my `node_modules` directory and I found that it works fine if a promise is returned, i.e.:

```js
if (!cancel) {
  return new Promise((resolve) => {
    timers[key] = setTimeout(function () {
      resolve(dispatch(action));
    }, time);
  })
}
```
I'm not sure if there are any other considerations to explicitly returning a promise in cases where `redux-thunk` is not in use, but this seems to work for my use case.
